### PR TITLE
drag face events

### PR DESCRIPTION
**Pull request by rezural**
- drag face input handling reworked to use events
- drag face selection input handling events wired up

This depends on #4, I will rebase if needed when you accept this one.

Cheers

### CODE REVIEW

File: src/edit_tools/drag_face.rs [hunk 1]
```diff
@@ -26,54 +30,36 @@ pub enum DragFaceState {
...
+pub enum DragFaceEvents {
+    StartDragFace(VoxelFace),
+    UpdateDragFace(Point3i),
+    FinishDragFace,
+}
+
+pub fn drag_face_default_input_map(
     voxel_cursor: VoxelCursor,
+    mut events: EventWriter<DragFaceEvents>,
+    mut current_tool: ResMut<CurrentTool>,
```
Comment (bonsairobo) [inline src/edit_tools/drag_face.rs:42]: I think this no longer needs to be `ResMut`.

File: src/edit_tools/drag_face.rs [hunk 2]
```diff
@@ -26,54 +30,36 @@ pub enum DragFaceState {
...
-                        controller.disable();
-                    }
-                    *state = DragFaceState::DraggingFace {
-                        quad_extent,
-                        normal,
-                        previous_drag_point: voxel_face.point,
-                    };
-                    *selection_state = SelectionState::Invisible;
-                }
+                events.send(DragFaceEvents::StartDragFace(voxel_face))
```
Comment (bonsairobo) [inline src/edit_tools/drag_face.rs:53]: I wouldn't send this event unless `quad_extent.contains(voxel_face.point)`. I.e. we're not actually going to start dragging if the selection was missed.

File: src/edit_tools/drag_face.rs [hunk 3]
```diff
@@ -26,54 +30,36 @@ pub enum DragFaceState {
...
-            let (quad_extent, normal) = if let SelectionState::SelectionReady {
-                quad_extent,
-                normal,
-            } = *selection_state
-            {
-                (quad_extent, normal)
-            } else {
-                return;
-            };
-
```
Comment (bonsairobo) [inline src/edit_tools/drag_face.rs:54]: I'd leave this here. We don't want to send events if there's not even a selection to drag yet.

File: src/edit_tools/drag_face.rs [hunk 4]
```diff
@@ -83,12 +69,69 @@ pub fn drag_face_tool_system(
...
+                    quad_extent,
+                    normal,
+                } = *selection_state
+                {
+                    (quad_extent, normal)
+                } else {
+                    return;
+                };
+
+                if quad_extent.contains(voxel_face.point) {
```
Comment (bonsairobo) [inline src/edit_tools/drag_face.rs:113]: All of this could stay in the input system that generates events.

**Review by bonsairobo — CHANGES_REQUESTED** [review]

File: src/edit_tools/selection/controller.rs [hunk 5]
```diff
@@ -21,49 +25,90 @@ pub enum SelectionState {
...
+            }
+            SelectionEvents::HoverMove(hover_face) => {
+                if let SelectionState::SelectingSecondCorner { first_corner, .. } = *selection_state
+                {
+                    if selection_corners_are_compatible(&first_corner, &hover_face) {
+                        let new_state = SelectionState::SelectingSecondCorner {
+                            first_corner,
+                            valid_hover: Some(*hover_face),
+                        };
+                        *selection_state = new_state;
```
Comment (bonsairobo) [inline src/edit_tools/selection/controller.rs:85]: No need for the `new_state` variable.

File: src/edit_tools/selection/controller.rs [hunk 6]
```diff
@@ -21,49 +25,90 @@ pub enum SelectionState {
...
+            SelectionEvents::SelectSecondCorner(hover_face) => {
+                if let SelectionState::SelectingSecondCorner {
+                    first_corner,
+                    valid_hover,
+                } = &mut *selection_state
+                {
+                    *valid_hover = None;
+                    // let mut valid_selection = false;
+                    if selection_corners_are_compatible(first_corner, &hover_face) {
+                        // valid_selection = true;
```
Comment (bonsairobo) [inline src/edit_tools/selection/controller.rs:98]: Remove the commented lines of code.

File: src/edit_tools/selection/controller.rs [hunk 7]
```diff
@@ -21,49 +25,90 @@ pub enum SelectionState {
...
+                    }
+                }
+            }
+            SelectionEvents::SelectSecondCorner(hover_face) => {
+                if let SelectionState::SelectingSecondCorner {
+                    first_corner,
+                    valid_hover,
+                } = &mut *selection_state
+                {
+                    *valid_hover = None;
```
Comment (bonsairobo) [inline src/edit_tools/selection/controller.rs:95]: This line isn't necessary right? In both cases below, we overwrite `selection_state` entirely.

**Review by bonsairobo — CHANGES_REQUESTED** [review]

File: src/edit_tools/drag_face.rs [hunk 8]
```diff
@@ -39,7 +39,8 @@ pub enum DragFaceEvents {
 pub fn drag_face_default_input_map(
     voxel_cursor: VoxelCursor,
     mut events: EventWriter<DragFaceEvents>,
-    mut current_tool: ResMut<CurrentTool>,
+    mut current_tool: Res<CurrentTool>,
+    mut selection_state: Res<SelectionState>,
```
Comment (bonsairobo) [inline src/edit_tools/drag_face.rs]: Doesn't need to be `mut`.

Comment (rezural) [reply src/edit_tools/drag_face.rs]: something weird going on, its not mut on my local copy.

**Pull request merged**